Overhaul code highlighting

Totto16 opened the pull request:
# Overhaul code highlighting 

## Description

As described in #271 code highlighting should become optional, so that user need to install an additional package or something similar to make it work.

There are a few options, I'll test all of them and list benefits from them and shortcomings


## Overview

We need an option that can do these things:
- detect languages, say how "confident" (highlight.js calls it relevance) the guess is, based on content only (no file extension or similar things, or shebangs)
- format languages to [Pango markup](https://docs.gtk.org/Pango/pango_markup.html) or something similar, that we can convert to it
- provide custom styles or use a set of predefined styles or be easy to parse / and or add custom styles to it
- support a few languages, not just 1,2


Our options are:

- 1: Use a command-line highlighter, that comes with many distributions, it can be installed and is then detected by Pano, to use it
- 2: use a JavaScript package as before, the user needs to install files into a correct location, and then Pano can detect it.
- 3: make a custom deb / rpm / arch etc. package with Pano and built-in highlighting, that installs the needed package automatically

### 1. Option

Pros: 
- the packages are quite robust and have many languages
- it is easy to install for the enduser.

Cons:
- we need to "shell out" e.g. spawn a shell process, that is more complex and error-prone than other methods


#### Feature Table

| package name | detection | many languages | markdown output | styleable |
|:--:|:-:|:-:|:-:|:-:|
|  [source-highlight](https://www.gnu.org/software/src-highlite/)  | ❌   | ✔️   | ⚠️    | ⚠️    |
|   [pygments](https://pygments.org/)|  ⚠️ |✔️   |   ✔️ | ✔️  |

#### Notes

`source-highlight`:

It can not detect languages without filename or shebang, it also can't tell you, how confident it was (there is a option to infer the language, but it only works on file extensions / shebangs or similar metadata)
It can't output to markdown, but to simple html, which would be styleable in some way
You can specify custom css or similar things, but it's not that easy, but maybe doable.


`pygments`(python3-pygments)

It is good at detecting creating languages (like python) but when I tried JS, or HTML or some other ones it failed ... i didn't test 100 files, but it wasn't that good, the one good thing was, that it's better than highlight.js in many cases and the detected languages all seem to look similar, when highlighted.
It has support for many many languages, it is easy styleable and can output pango markdown out of the box.

update: after some modification to the detection logic, it seems really stable and robust when working with more than ~ +- 200 characters. The shortcoming of it is, that it has no real relevance value, but we can emulate that, by using that charaacter length. After that it is really good.




### 2. Option

Pros:
- it is easy to integrate
- it has many options 
- many highlighter have many languages and are quite good


Cons:
- If we don't ship them, it's not easy for the end-user to install



### 3. Option


Pros:
- We can pre package everything we need, the cons of the 2. Option or the previous implementation go away
- We can select option 1 or 2 as core implementation

Cons:
- We would stop publishing on EGO or the package published there is not complete and maybe, depending on the use option (1 or 2) it might be not easy to add the code highlighting to Pano again by hand


## Current state

I would focus on option 1, since it's the best choice atm. I tested `source-highlight` but it's quite bad. So I'll be using `pygments`(python3-pygments) for now.

TODOs:

- [ ] fix all bugs with the new implementation
- [ ] fix slider (Gtk4.Scale) Size in Settings
![grafik](https://github.com/oae/gnome-shell-pano/assets/32566573/b2b64bb5-f1b6-4031-a7bf-18b29737ff26)
- [ ] more testing
- [ ] fix search of styles in settings

optional:
- [ ] Improve pano theme for pygments





## Type of change


- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] My commits follow the commit standards of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings

Totto16 committed:
remove highlight.js

Totto16 committed:
remove prismjs, as it is not needed anymore (atm)

Totto16 committed:
implement code highlighting interface,, remove prismjs and highlight.js references

Totto16 committed:
fix errors an make highlighting work as intended, when it is disabled

Totto16 commented:
@oae Do you have an auto branch protection rule in some way, if the branch isn't named `feat/` (is that even possible)
Since I am currently unable to push (it fails with the reason, that this is a protected branch ❓) 
![grafik](https://github.com/oae/gnome-shell-pano/assets/32566573/408a237e-2d71-4e86-a21f-1921c4c810f7)

oae commented:
The first option with pygments sounds good to me too

oae commented:
> @oae Do you have an auto branch protection rule in some way, if the branch isn't named `feat/` (is that even possible) Since I am currently unable to push (it fails with the reason, that this is a protected branch ❓) ![grafik](https://private-user-images.githubusercontent.com/32566573/324208926-408a237e-2d71-4e86-a21f-1921c4c810f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM2NTU3NzksIm5iZiI6MTcxMzY1NTQ3OSwicGF0aCI6Ii8zMjU2NjU3My8zMjQyMDg5MjYtNDA4YTIzN2UtMmQ3MS00ZTg2LWEyMWYtMTkyMWM0YzgxMGY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIwVDIzMjQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwNmQ4YWY3MDhhNTFlZDgxZThiMjllZjZlNTI1MTdkNDRmMTU2ODQ1NjQxNDM5ZDI4NWE2NWYxMWZlOTg3YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.v2Ow_86RXUB3YWo77K3ScDsnTsrEqy8s13Amnn3PCHg)

I have changed some settings now. Can you try again?

Totto16 commented:
> The first option with pygments sounds good to me too



> > @oae Do you have an auto branch protection rule in some way, if the branch isn't named `feat/` (is that even possible) Since I am currently unable to push (it fails with the reason, that this is a protected branch ❓) ![grafik](https://private-user-images.githubusercontent.com/32566573/324208926-408a237e-2d71-4e86-a21f-1921c4c810f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM2NTU3NzksIm5iZiI6MTcxMzY1NTQ3OSwicGF0aCI6Ii8zMjU2NjU3My8zMjQyMDg5MjYtNDA4YTIzN2UtMmQ3MS00ZTg2LWEyMWYtMTkyMWM0YzgxMGY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIwVDIzMjQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwNmQ4YWY3MDhhNTFlZDgxZThiMjllZjZlNTI1MTdkNDRmMTU2ODQ1NjQxNDM5ZDI4NWE2NWYxMWZlOTg3YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.v2Ow_86RXUB3YWo77K3ScDsnTsrEqy8s13Amnn3PCHg)
> 
> I have changed some settings now. Can you try again?

Same error, strange

oae commented:
I think it should be fixed now

Totto16 committed:
implement basic code highlighting with pygments
- use interface that permits multiple highlighters in the future

Totto16 committed:
WIP: add settings for code highlighter

Totto16 committed:
Code Highlighter Settings:
- fix many settings errors, so that they behave correctly
- use setter of `sensitive` instead of `set_sensitive`
- correctly connect settings and the core `_markdownDetector` of the extension
- display code items as text items, if there is no code highlighter, or he is disabled

Totto16 committed:
improve eslint config

Totto16 committed:
use async subprocess calls

Totto16 committed:
- fix all eslint errors /warnings
- change paths, that now require async routes in the code flow
- add LoadingPanoItem, that shows up, before an item is loaded

Totto16 committed:
- remove the loadingPanoItem and deal with the async highlighting in another way:
 make an not highlighted version of the code item, and after the highlighting returns, set the correct markdown
 - type metaData of all items explicitly with new TS wrapper methods to JSOn.parse and JSON.stringify

Totto16 committed:
fix updating of codePanoItem, so that it works as expected

Totto16 committed:
fix a few async errors:
- always catch errors
- add load status to PangoMarkdown

Totto16 committed:
format eslintrc

Totto16 committed:
use newer "date-fns" package

Totto16 committed:
update rollup

Totto16 committed:
many fixes:
- add mor logger types and use them
- add a slider(Gtk4.Scale) to set the relevance threshold for the highlighter
- fix Promisified types
- add some more inline comments about (not needed atm) not implemented features
- chnage CodeItem handling of later initialization

Totto16 committed:
fix initial creation of code items, it now waits, until the markdownDetector is loaded

Totto16 committed:
fix scale row

Totto16 committed:
- fix log spam in case of disabled highlighter

Totto16 committed:
remove scan method and use onLoad of markdownDetector to get asynchronous scanning
- fix createDropdownRowForHighlighter, it had some errors in using some outer scope variables incorrectly

Totto16 committed:
fix a few errors:
- fix scale issue, when resetting
- fix small issue in panoItemFactor
- refactor how teh style in codePanoItem is handled, it now also listens to text events and can mimic text style, only if it's set to code it display the code

Totto16 committed:
- add mechanism to schedule concurrent markupCode request, so that in the beginning and when changing theme it doesn't create 100 of shell requests in one thread, this might slow down the gnome-shell, so distributing it is better

Totto16 committed:
add custom pygments style

Totto16 committed:
use poetry and better python pygments style config

Totto16 committed:
- completely add the python styles
- use it by setting PYTHONPATH env variable to the correct place
- installing it into the dist folder (might be a bad idea)

Totto16 committed:
minor dependency updates

Totto16 committed:
add a note about the validate-color package for the future

Totto16 committed:
catch exception, that might occur (it's really rare, and in the debug environment only)

Totto16 committed:
- don't build the python pygments wheel by default, but still offer a way to build it
- add Pipenv, to install poetry, for the build step
- change the directory, where it's locally installed (it's optional now)
- set the PYTHONPATH nevertheless, since if it's not there, it doesn't matter, later we can offer the wheel for users to install globally, so that it's opt-in (we porblay can also publish it to pypi?

Totto16 committed:
- also test the pygments style plugin build in teh CI
- change the PYTHONPATH location in the code

Totto16 committed:
- prettier: ignore python code

Totto16 committed:
- add some void annotation to potential promises, that we ignore (it is intended like that)

Totto16 committed:
- python ci, try to fix it
- pygments styles, add more colors and bump version

Totto16 committed:
- sort style values of highlighter
- add more sophisticated scheduled job handling: now they can be cancelled, which is done at disabling and also, if we disable the code highlighter, so that no unnecessary tasks are run

Totto16 committed:
use  GLib.timeout_add instead of setTimeout

Totto16 committed:
fix ci

Totto16 committed:
- refactor constructor of PangoMarkdown, so that each constructed Highlighter gets it's options at the start from teh settings
- add SpinRow for the Highlighter settings
- add relevanceLength setting to pygments, taht replaces a hardcoded value

Totto16 committed:
fix small dumb mistake

Totto16 committed:
bump dependencies

Totto16 committed:
- update @girs type dependencies
- completely reinstall the dependencies, to get newer sub-dependencies

Totto16 committed:
fix property name change in gnome-shell types

HearseDev commented:
Not really experienced with anything gnome, but was wondering if this would be a viable and much cleaner solution. It could be an optional lib install and also has bindings for js. Or it could be required and could be used as a default text view? I am just throwing out an idea, never really worked with anything remotely close to this.
https://wiki.gnome.org/Projects/GtkSourceView

Totto16 commented:
> Not really experienced with anything gnome, but was wondering if this would be a viable and much cleaner solution. It could be an optional lib install and also has bindings for js. Or it could be required and could be used as a default text view?(think it might be preinstalled with gnome, not too sure) I am just throwing out an idea, never really worked with anything remotely close to this. https://wiki.gnome.org/Projects/GtkSourceView https://gnome.pages.gitlab.gnome.org/gtksourceview/gtksourceview5/class.PrintCompositor.html

I think I saw this in my research, but It isn't half as good as [pygments](https://pygments.org/) in autodetecting languages and it's ballpark is nearly the same as [source-highlight](https://www.gnu.org/software/src-highlite/).

It has integrated GTK support, but it's not themeable (at least easily) and it has no "real" autodetection feature, you have to create a widget to get the language, which is not suitable for our purpose 😓

PiotrGrobelak commented:
Hi @oae and @Totto16, what can I pick up to help you with for the full release of GNOME 46?

Malix-Labs commented:
I don't know if it's relevant, but have you thought about using Deno instead of Node ?

It would help drastically reduce the number of dependencies required by EGO

Totto16 commented:
> I don't know if it's relevant, but have you thought about using Deno instead of Node ?
> 
> It would help drastically reduce the number of dependencies required by EGO

We don't use node.js, this is a gnome extension, so it uses [GJS](https://gitlab.gnome.org/GNOME/gjs)

Malix-Labs commented:
Thanks for the clarification

Also, have you noticed the kind support offered by @PiotrGrobelak : https://github.com/oae/gnome-shell-pano/pull/278#issuecomment-2171365881 ?

Totto16 commented:
> Thanks for the clarification
> 
> Also, have you noticed the kind support offered by @PiotrGrobelak : [#278 (comment)](https://github.com/oae/gnome-shell-pano/pull/278#issuecomment-2171365881) ?

I did, but this I rather complicated and I have to clean a few things up 😓 
It is not easy to help here, as this is a complicated thing to do, since it requires some really complex things to do, where you have to be familiar with many things regarding pano and how they work internally 😓

Malix-Labs commented:
Understandable

So only you @Totto16 and @oae could ship this PR and make this extension mainstream again ?

teohhanhui commented:
I think option 2 is viable.

We'll just have to instruct the user to do something like:

```
sudo dnf install nodejs
cd ~/.local/share/pano@elhan.io
npm install highlight.js@^11.9.0
```

Malix-Labs commented:
I strongly advocate against using nodejs as a dependency, which would mean Pano would rely on a different runtime that the one shipped with GNOME (GJS, apparently)

teohhanhui commented:
> I strongly advocate against using nodejs as a dependency, which would mean Pano would rely on a different runtime that the one shipped with GNOME (GJS, apparently)

That's not the case at all. It's just a way for the user to install the package from npm.

Malix-Labs commented:
> it basically would mean to rely on a third party (APT, RPM, nixpkgs, etc…) third party (npm) package manager

> Dropping support for highlighting for now seems like a way more acceptable option

Malix-Labs commented:
The optional part would be understandable indeed, but it would probably be simpler to advice to use the non-EGO release for added features instead

A rewrite using something else than highlight would be optimal, for the future (but would take time)

teohhanhui commented:
> A rewrite using something else than highlight would be optimal, for the future (but would take time)

Honestly I don't understand this position. It seems to me that the problem is EGO's review process. But it is how it is.

Malix-Labs commented:
The problem is losing support for EGO which transitively means losing an integrated GNOME experience (install, update, etc)

teohhanhui commented:
> The problem is specifically losing support for EGO which transitively means losing an integrated GNOME experience (install, update, etc)

Absolutely agreed.

However

> but it would probably be simpler (both for the end-user and for the docs team) to ship a non-EGO release with dependencies for the added features instead

This suffers from the problem of no auto updates as well. :disappointed:

Malix-Labs commented:
> many dependencies

To my knowledge, only `libgda` and `libgda-sqlite`, which are sometimes already shipped by default in some Linux distributions, and are both combined nowhere close to node + highlightjs dependency bundle size

> This suffers from the problem of no auto updates as well. 😞

This is correct, yep

Don't you think that, for the long run, it would be preferable to rewrite to a more system-agnostic + EGO compliant extension ?

In the meantime, documenting to add node and highlights as optional dependencies (with a size warning) could be a viable option ; The other being dropping total support for highlighting

teohhanhui commented:
> Don't you think that, for the long run, it would be preferable to rewrite to a more system-agnostic + EGO compliant extension ?

Option 1 also does not do that. The user would have to manually install another system dependency (`highlight`) through their package manager.

I don't think there are any realistic good options with the current EGO stance...

(Personally I don't care about code highlighting. I don't even use the URL preview because it's a major security and privacy risk.)

teohhanhui commented:
For option 2, there's also `tree-sitter highlight`: https://tree-sitter.github.io/tree-sitter/syntax-highlighting

ArthurValadares commented:
Correct me if I'm wrong, but the EGO Guideline refers to entire applications, in order to facilitate the review process and not add unnecessary tools for the user. Would a library for accessing information available on the Internet with the user's authorisation fall under this heading? 

Highlight being disabled by default is the right way to go, but for the user to enable it manually would add a complicated step for many. In order to maintain the structure of the project, wouldn't it be prudent to add a more modular character to highlight - prismjs too, if appropriate - and add the package [pacote](https://www.npmjs.com/package/pacote) to the pano dependencies and through it, with the user's authorisation, download and enable highlight?
